docs: DOC-205: Document changes to ML backend (#5642)
heartexlabs/label-studio: docs/source/guide/ml.md

section: "Machine learning"
---

You can use an ML backend to integrate your model development pipeline with your data labeling workflow. There are several use cases, including: 

> - Pre-annotate data with a model

Reviewer: Pre-annotate, or Autolabel data with model

- Use active learning to select the most relevant data for labeling
- Interactive (AI-assisted) labeling
- Model fine-tuning based on recently annotated data

For example, for an image classification task, the model pre-selects an image class for data annotators to verify. For audio transcriptions, the model displays a transcription that data annotators can modify. 

!!! info Tip
    You can use [Label Studio Enterprise to build an automated active learning loop](https://docs.humansignal.com/guide/active_learning.html) with a machine learning model backend. If you use the open source Community Edition of Label Studio, you can manually sort tasks and retrieve predictions to mimic an active learning process.

> If you just need to load static pre-annotated data into Label Studio, running an ML backend might be overkill for you. Instead, you can [import preannotated data](predictions).

Reviewer: I feel before this line, we have to explain how it works:
1. User opens the task
2. LS sends the request to ML backend
3. ML backend return the online prediction response
4. LS shows prelabeling on the screen

Author: See line 21 - added list

## Set up an example ML backend

The Label Studio ML backend is an SDK that wraps your machine learning code and turns it into a web server. The web server can be connected to a running Label Studio instance to automate labeling tasks. We have provided a [library of example models](https://github.com/HumanSignal/label-studio-ml-backend?tab=readme-ov-file#models) that you can use in your own workflow, or extend and customize as needed. 

Ir you want to write your own model instead, see [Write your own ML backend](ml_create).

### Prerequisites

<div class="opensource-only">

* [Label Studio](install)
* [Docker Compose](https://docs.docker.com/compose/install/)

</div>

<div class="enterprise-only">

* [Label Studio](install_enterprise)
* [Docker Compose](https://docs.docker.com/compose/install/)

</div>

### Start the model

1. First, decide which [model](#Example-models) you want to use and check for required parameters (click the link for each model to see a full parameter list). 

>     Use [`label_studio_ml/default_configs/docker-compose.yml`](https://github.com/HumanSignal/label-studio-ml-backend/blob/master/label_studio_ml/default_configs/docker-compose.yml) for general configuration, and `docker-compose.yml` within the model directory for model-specific parameters. 

Reviewer: we don't need to refer general configuration here - always use `docker-compose.yml` from the model directory, it contains all configuration options.
You only have to know about `default_configs` when you start creating your own ML backend

2. Then replace `{MODEL_NAME}` in the below command with the appropriate directory. 

>     For example, if you are using the SAM backend, the model name would be `segment_anything_model`, which matches the [directory name in the repo](https://github.com/HumanSignal/label-studio-ml-backend/tree/master/label_studio_ml/examples/segment_anything_model):

Reviewer: ```suggestion
    For example, if you are using the SAM backend with [SegmentAnything model](https://segment-anything.com/), the model name would be `segment_anything_model`, which matches the [directory name in the repo](https://github.com/HumanSignal/label-studio-ml-backend/tree/master/label_studio_ml/examples/segment_anything_model):
```

    ```bash
git clone https://github.com/HumanSignal/label-studio-ml-backend.git

> cd label-studio-ml-backend/label_studio_ml/examples/{MODEL_NAME}

Reviewer: ```suggestion
cd label-studio-ml-backend/label_studio_ml/examples/segment_anything_model
```

docker-compose up
    ```

The model should begin running at `http://localhost:9090`. You can verify this by clicking **Send Test Request** from the overflow menu next to the model or by using the following command: 

> `curl http://localhost:9090/health`

Reviewer: ```suggestion
```bash
> curl http://localhost:9090
{"model_class":"SamMLBackend","status":"UP"}
```
The healthy output contains `"status"` field and the running model name
```

If you see any errors, see [Troubleshooting ML Backends & Predictions](https://support.humansignal.com/hc/en-us/sections/23627938255117-ML-Backend-Predictions) in the HumanSignal support center. 

### Connect the model to Label Studio

<div class="enterprise-only">

After you [create a project](setup_project), open the project settings and select [**Model**](project_settings_lse#Model). 

</div>

<div class="opensource-only">

After you [create a project](setup_project), open the project settings and select [**Model**](project_settings#Model). 

</div>

Click **Connect Model** and complete the following fields:

| Field | Description                                                                            |
| -------- | -------------------------------------------------------------------------------------- |
| **Name**   | Enter a name for the model.                        |

> | **Backend URL**  | Enter a URL for the model. <br /><br />If you are following the steps above, this would be `http://localhost:9090`.  |

Reviewer: We should add the note about `localhost` can be inaccessible if LS is running from docker (as it has been mentioned above)

| **Select authentication method**   | If a username and password are required to access the model, you can select **Basic Authentication** and enter them here.                                     |
| **Extra params**  | Enter any additional parameters you want to pass to the model.                                      |

> | **Interactive preannotations**  | Enable this option to allow the model to assist with the labeling process by providing real-time predictions or suggestions as annotators work on tasks.  <br /><br />In other words, as you interact with data (for example, by drawing a region on an image, highlighting text, or asking an LLM a question), the ML backend receives this input and returns predictions based on it.  |

Reviewer: should be the link to interactive labeling guide here

!!! info Tip
    You can also [add an ML backend using the API](/api/#operation/api_ml_create). You will need the project ID and the machine learning backend URL. 

## Example models

> For a list of the available example models, see the [Models table in GitHub](https://github.com/HumanSignal/label-studio-ml-backend?tab=readme-ov-file#models). 

Reviewer: Why don't we want to move the full table here for better visibility? or is it planned for the future?

Author: I had that table here, but then you kept updating it! I can add it back once it's more finalized.

Reviewer: it is finalized now

Author: Added table

Some of them work without any additional configuration. If the model has required parameters, you can set those parameters in `docker-compose.yml` within the model directory. 

## Model training

Training a model allows it to learn from submitted annotations and potentially improve its predictions for subsequent tasks. 

After you connect a model to Label Studio as a machine learning backend and annotate at least one task, you can start training the model. You can use automated or manual training. 

<div class="enterprise-only">

From the [**Model** page](project_settings_lse#Model) under project settings, select one of the following: 

</div>

<div class="opensource-only">

From the [**Model** page](project_settings#Model) under project settings, select one of the following: 

</div>

> * **Start model training on annotation submission**--Enable this option for automated training. When enabled, training is automatically initiated every time an annotation is submitted or updated. 

Reviewer: Please remove this line

> * **Start Training** (Available from the overflow menu next to the connected model)--Manually initiate training. Use this action if you want to control when the model training occurs, such as after a specific number of annotations have been collected or at certain intervals.

Reviewer: ```suggestion
* **Start Training** (Available from the overflow menu next to the connected model) -- Manually initiate training. Use this action if you want to control when the model training occurs, such as after a specific number of annotations have been collected or at certain intervals.
```

You can also initiate training programmatically using the following:

* From the API, specify the ID of the machine learning backend and run the following command: 
   ```
   curl -X POST http://localhost:8080/api/ml/{id}/train
   ```
  See [the Train API documentation](/api/#operation/api_ml_train_create) for more.
- [Trigger training with webhooks](ml_create#Trigger-training-with-webhooks).

> In development mode, training logs appear in the web browser console. In production mode, you can find runtime logs in `my_backend/logs/uwsgi.log` and RQ training logs in `my_backend/logs/rq.log` on the server running the ML backend, which might be different from the Label Studio server. 

Reviewer: it is obsolete - no wsgi/rq processes exist. All training logs stream in stdout / console

To see more detailed logs, start the ML backend server with the `--debug` option. 

## Pre-annotations/predictions

!!! note
    The terms "predictions" and pre-annotations" are used interchangeably. 

### Get predictions from a model

After you connect a model to Label Studio, you can see model predictions in the labeling interface if the model is pre-trained, or right after it finishes [training](#Model-training). 

* To manually add predictions, go to the Data Manager, select the tasks you want to get predictions for, and then select **Actions > Retrieve predictions**.

> * To automatically pre-label data with predictions, go to the project settings and enable **Annotation > Use predictions to prelabel data**. 

Reviewer: we should add the note:
"ensure the connected model is selected in dropdown `Select which predictions or which model you want to use:`"

  ]
}
```

### Interactive pre-annotations

ML-assisted labeling with interactive pre-annotations works with image segmentation and object detection tasks using rectangles, ellipses, polygons, brush masks, and keypoints, as well as with HTML and text named entity recognition tasks. Your ML backend must support the type of labeling that you're performing, recognize the input that you create, and be able to respond with the relevant output for a prediction.

Either enable the **Interactive preannotations** option when adding a model, or use the **Edit** action from the project settings page to enable/disable this option. 

#### Smart tools

> For image labeling, you can update your labeling configuration to include the `smart="true"` option for the type of labeling you're performing. Smart tools appear by default if auto-annotation is enabled. If you only want the smart option to appear and don't want to perform manual labeling at all, use `smartOnly="true"`. 

Reviewer: It seems we need to elaborate on what smart tools are:
- tools that duplicate controls to allow interaction with connected ML backend (currently supported for <RectangleLabels>, <Keypoints>, <BrushLabels>, <Ellipse>, <POlygon>)
- give the data in `context` variable in ML backend /predict request + link to explanation what the context variable is and how to use it

  ]
}
```

### Interactive pre-annotations

ML-assisted labeling with interactive pre-annotations works with image segmentation and object detection tasks using rectangles, ellipses, polygons, brush masks, and keypoints, as well as with HTML and text named entity recognition tasks. Your ML backend must support the type of labeling that you're performing, recognize the input that you create, and be able to respond with the relevant output for a prediction.

Either enable the **Interactive preannotations** option when adding a model, or use the **Edit** action from the project settings page to enable/disable this option. 

#### Smart tools

> Smart tools duplicate the tools that you have set up to use through your labeling configuration (e.g. rectangles) to allow interaction with the ML backend. Smart tools are currently for Rectangle, Ellipse, Polygon, Keypoint, and Brush tags. 

Reviewer: shall we provide exact labeling tags instead?

Author: Added tag formatting and links to the appropriate tag pages

> Smart tools are dynamic and can change behavior or appearance based on the context, leveraging machine learning predictions. For example, smart tools can automatically detect and suggest annotations for objects in images, which annotators can then review and refine. This can significantly speed up the labeling process, especially when dealing with large datasets.

Author: Definitely read through this new section

> Smart tools use `context` to adapt their behavior based on the current state of the labeling environment and the specific needs of the task at hand. For instance, they might alter their functionality depending on the type of data being labeled or the particular region of interest within a task. W
> 
> hen smart tools are enabled for a labeling task, they can interact with the ML backend by sending data to the `/predict` endpoint to receive predictions. These predictions are then used to provide interactive pre-annotations within the labeling interface. For more information, see [Support interactive pre-annotations in your ML backend](ml_create#Support-interactive-pre-annotations-in-your-ML-backend). 

Reviewer: ```suggestion
Smart tools use `context` to adapt their behavior based on the current state of the labeling environment and the specific needs of the task at hand. For instance, they might alter their functionality depending on the type of data being labeled or the particular region of interest within a task. When smart tools are enabled for a labeling task, they can interact with the ML backend by sending data to the `/predict` endpoint to receive predictions. These predictions are then used to provide interactive pre-annotations within the labeling interface. For more information, see [Support interactive pre-annotations in your ML backend](ml_create#Support-interactive-pre-annotations-in-your-ML-backend). 
```

### Choose which predictions to display to annotators

<div class="opensource-only">

> You can choose which model or prediction set to display to annotators by default. This is available under the **Annotation > Pre-labeing** in the project settings. 

Reviewer: ```suggestion
You can choose which model or prediction set to display to annotators by default. This is available under the **Annotation > Live Predictions** in the project settings. 
```

</div>

<div class="enterprise-only">

> You can choose which model or prediction set to display to annotators by default. This is available under the **Annotation > Predictions** in the project settings. 

Reviewer: ```suggestion
You can choose which model or prediction set to display to annotators by default. This is available under the **Annotation > Live Predictions** in the project settings. 
```
